### PR TITLE
Use allocator directly instead of allocator_traits

### PR DESCRIPTION
**Pull request by feresr**
Hi! First of all, thanks for this amazing resource. I've made a [Super Mario Bros clone](https://github.com/feresr/super-mario-bros) using this library. 

I didn't use your library as-is but made my own dumbed down version based on yours (mainly because I wanted to understand 100% of the code in my game, I'm new to C++). 

Now that the game is "complete" I want to fully understand your code in order to learn more advanced c++. 
I noticed you are making heavy use of Allocators (which I'm having a really hard time wrapping my head around). 
I was wondering why you use this "allocator_traits" function to access the allocator instead of just using the allocator you already defined.

If you have any reference on where I can learn more about allocators I would greatly appreciate it. Feel free to close this PR as I'm sure doesn't improve the code at all.

### CODE REVIEW

**Comment by redxdev** [comment]
I wrote this a while ago and now there's lots I'd change (mostly relating to where things are in memory), but I'm glad people are still finding it useful!

`allocator_traits` is a somewhat unfortunate side-effect of C++'s ecosystem. It isn't necessary if every type of allocator matches the exact interface of those available in the STL, but there are lots of third-party libraries that might not have the exact same methods and fields available for allocators they provide.

By using [template specialization](https://en.cppreference.com/w/cpp/language/template_specialization) you can change the behavior of `allocator_traits` depending on what type of allocator is being used. In that way you can use an allocator that otherwise wouldn't be usable with classes expecting an STL allocator. There's more info on `allocator_traits` [here](https://en.cppreference.com/w/cpp/memory/allocator_traits).

tldr: it makes compatibility with non-STL allocators easier, but isn't strictly necessary.

**Comment by feresr** [comment]
Ohh I see, it's like the C++ equivalent of a Java interface (kindof..)? Thanks for getting back to me!

**Comment by redxdev** [comment]
Sort of.

C++ already has a better equivalent of an interface from Java/C#/etc. It's just a normal class with only pure virtual functions:

```cpp
class MyInterface
{
public:
  virtual void DoSomething() = 0;
  virtual void DoSomethingElse() = 0;
};

class MyImplementingClass : public MyInterface
{
public:
  virtual void DoSomething() override { /* ... */ }
  virtual void DoSomethingElse() override { /* ... */ }
};
```

You could theoretically have an "IAllocator" interface and then either implement that for an allocator or use it to forward to a non-standard implementation of an allocator.

The problem with the above is that it isn't "free" - virtual functions have a performance cost associated with them since figuring out which implementation gets called is done at runtime (see [vtables](https://pabloariasal.github.io/2017/06/10/understanding-virtual-tables/)).

What's neat about what `allocator_traits` does is that template specialization is entirely figured out at compile-time - calling methods via `allocator_traits` is calling a non-virtual method, but the exact method that gets called is dependent on the template type.

Template specialization can't really replace interfaces in most cases because the exact type of the template _must_ be known in full at compile-time, but since you very rarely (if ever) would want to change what allocator you're using at runtime for a specific object that isn't an issue here.

**Comment by feresr** [comment]
Thanks for the clear example, this makes a lot of sense.
I've read about vtables and the advantages(and disadvantages) of c++ `templates` over things like `interfaces` and `generics` in languages Java/C#/Kotlin but never seen those in practice. 

It's all starting to click for me. Thanks again for all the help!
